File: clients/scala/src/main/scala/ai_cup_22/model/Zone.scala
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Current state of the game zone
 *
 * @param currentCenter Current center
 * @param currentRadius Current radius
 * @param nextCenter Next center
 * @param nextRadius Next radius
 */
case class Zone(currentCenter: ai_cup_22.model.Vec2, currentRadius: Double, nextCenter: ai_cup_22.model.Vec2, nextRadius: Double) {
    /**
     * Write Zone to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        currentCenter.writeTo(stream)
        StreamUtil.writeDouble(stream, currentRadius)
        nextCenter.writeTo(stream)
        StreamUtil.writeDouble(stream, nextRadius)
    }

    /**
     * Get string representation of Zone
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Zone { ")
        stringBuilder.append("currentCenter: ")
        stringBuilder.append(currentCenter)
        stringBuilder.append(", ")
        stringBuilder.append("currentRadius: ")
        stringBuilder.append(currentRadius)
        stringBuilder.append(", ")
        stringBuilder.append("nextCenter: ")
        stringBuilder.append(nextCenter)
        stringBuilder.append(", ")
        stringBuilder.append("nextRadius: ")
        stringBuilder.append(nextRadius)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Zone {
    /**
     * Read Zone from input stream
     */
    def readFrom(stream: java.io.InputStream): Zone = Zone(
        ai_cup_22.model.Vec2.readFrom(stream),
        StreamUtil.readDouble(stream),
        ai_cup_22.model.Vec2.readFrom(stream),
        StreamUtil.readDouble(stream)
    )
}

File: clients/swift/Sources/AiCup22/Model/Zone.swift
/// Current state of the game zone
public struct Zone {
    /// Current center
    let currentCenter: Vec2

    /// Current radius
    let currentRadius: Double

    /// Next center
    let nextCenter: Vec2

    /// Next radius
    let nextRadius: Double

    /// Read Zone from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Zone {
        var currentCenter: Vec2
        currentCenter = Vec2.readFrom(stream)
        var currentRadius: Double
        currentRadius = stream.readDouble()
        var nextCenter: Vec2
        nextCenter = Vec2.readFrom(stream)
        var nextRadius: Double
        nextRadius = stream.readDouble()
        return Zone(currentCenter: currentCenter, currentRadius: currentRadius, nextCenter: nextCenter, nextRadius: nextRadius)
    }

    /// Write Zone to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        currentCenter.writeTo(stream)
        stream.writeDouble(currentRadius)
        nextCenter.writeTo(stream)
        stream.writeDouble(nextRadius)
    }
}
